File: readfq.py
def readfq(fp): # this is a generator function
    import string
    finished, last = False, None
    while (True): # mimic closure; is it a bad idea?
        if (finished): raise StopIteration # already hit the end-of-file (EOF)
        if (last is None): # the first record or a record following a fastq
            for l in fp: # search for the start of the next record
                if l[0] in '>@':
                    last = l[:-1] # save this line
                    break
        if last is None: # EOF
            finished = True
            raise StopIteration
        name = last[1:].split()[0]
        seqs, last, c = [], None, None
        for l in fp: # read the sequence
            c = l[0]
            if c in "@+>":
                last = l[:-1]
                break
            seqs.append(l[:-1])
        if (last is None): finished = True # EOF
        if (c != '+'): yield name, ''.join(seqs), None # yield a fasta record
        else: # this is a fastq record
            seq, leng, seqs = ''.join(seqs), 0, []
            for l in fp: # read the quality
                seqs.append(l[:-1])
                leng += len(l) - 1
                if (leng >= len(seq)): # have read enough quality
                    last = None
                    yield name, seq, ''.join(seqs); # yield a fastq record
                    break
            if (last): # reach EOF before reading enough quality
                finished = True
                yield name, seq, None # yield a fasta record instead

if __name__ == "__main__":
    import sys
    n, slen, qlen = 0, 0, 0
    for name, seq, qual in readfq(sys.stdin):
        n += 1
        slen += len(seq)
        qlen += qual and len(qual) or 0
    print n, '\t', slen, '\t', qlen

File: readfq.lua
local function readfq(fp)
	local finished, last = false, nil;
	return function() -- closure
		if finished then return nil end
		if last == nil then -- the first record or a record following a fastq
			for l in fp:lines() do
				if l:byte(1) == 62 or l:byte(1) == 64 then -- ">" || "@"
					last = l;
					break;
				end
			end
			if last == nil then -- reach the end of the file (EOF)
				finished = true;
				return nil;
			end
		end
		local i = last:find("%s"); -- separate the sequence name and the comment
		name, last = i and last:sub(2, i-1) or last:sub(2), nil; -- sequence name
		local seqs, c = {}; -- c is the first character of the last line
		for l in fp:lines() do -- read sequence
			c = l:byte(1);
			if c == 62 or c == 64 or c == 43 then -- ">" || "@" || "+"
				last = l;
				break;
			end
			table.insert(seqs, l); -- similar to python, string cat is inefficient
		end
		if last == nil then finished = true end -- end of file
		if c ~= 43 then return name, table.concat(seqs) end -- a fasta record
		local seq, len = table.concat(seqs), 0; -- prepare to parse quality
		seqs = {};
		for l in fp:lines() do -- read quality
			table.insert(seqs, l);
			len = len + #l;
			if len >= #seq then -- we have read enough qualities
				last = nil;
				return name, seq, table.concat(seqs); -- a fastq record
			end
		end
		finished = true; -- reach EOF
		return name, seq; -- incomplete fastq quality; return a fasta record
	end
end

-- for testing only

local n, slen, qlen = 0, 0, 0
for name, seq, qual in readfq(io.stdin) do
	n, slen = n + 1, slen + #seq
	qlen = qlen + (qual and #qual or 0)
end
print(n, slen, qlen)
